ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/UnifiedProtocolSchedule.java

```diff
@@ -0,0 +1,218 @@
+/*
+ *
+ *  * Copyright Hyperledger Besu Contributors.
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ *  * the License. You may obtain a copy of the License at
+ *  *
+ *  * http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ *  * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *  * specific language governing permissions and limitations under the License.
+ *  *
+ *  * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package org.hyperledger.besu.ethereum.mainnet;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.stream.Collectors.partitioningBy;
+
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.ProcessableBlockHeader;
+import org.hyperledger.besu.ethereum.core.TransactionFilter;
+import org.hyperledger.besu.ethereum.worldstate.WorldStateArchive;
+
+import java.math.BigInteger;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableSet;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.TreeSet;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.common.annotations.VisibleForTesting;
+
+public class UnifiedProtocolSchedule implements ProtocolSchedule {
+
+  @VisibleForTesting
+  protected NavigableSet<ScheduledProtocolSpec> protocolSpecs =
+      new TreeSet<>(Comparator.comparing(ScheduledProtocolSpec::milestone).reversed());
+
+  private final Optional<BigInteger> chainId;
+
+  public UnifiedProtocolSchedule(final Optional<BigInteger> chainId) {
+    this.chainId = chainId;
+  }
+
+  @VisibleForTesting
+  protected UnifiedProtocolSchedule(final UnifiedProtocolSchedule protocolSchedule) {
+    this.chainId = protocolSchedule.chainId;
+    this.protocolSpecs = protocolSchedule.protocolSpecs;
+  }
+
+  @Override
+  public ProtocolSpec getByBlockHeader(final ProcessableBlockHeader blockHeader) {
+    //    checkArgument(number >= 0, "number must be non-negative");
```

Comment:
Think we can safely assume that a blockHeader will never have a negative number or timestamp